exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/ResourceMarshaler.java

```diff
@@ -16,41 +16,62 @@
   private static final WeakConcurrentMap<io.opentelemetry.sdk.resources.Resource, ResourceMarshaler>
       RESOURCE_MARSHALER_CACHE = new WeakConcurrentMap.WithInlinedExpunction<>();
 
-  private final byte[] serializedResource;
+  private final byte[] serializedBinary;
+  private final String serializedJson;
 
   static ResourceMarshaler create(io.opentelemetry.sdk.resources.Resource resource) {
     ResourceMarshaler cached = RESOURCE_MARSHALER_CACHE.get(resource);
     if (cached == null) {
       // Since WeakConcurrentMap doesn't support computeIfAbsent, we may end up doing the conversion
       // a few times until the cache gets filled which is fine.
-      cached = new ResourceMarshaler(KeyValueMarshaler.createRepeated(resource.getAttributes()));
+
+      RealResourceMarshaler realMarshaler =
+          new RealResourceMarshaler(KeyValueMarshaler.createRepeated(resource.getAttributes()));
+
+      ByteArrayOutputStream binaryBos =
+          new ByteArrayOutputStream(realMarshaler.getBinarySerializedSize());
+
+      try {
+        realMarshaler.writeBinaryTo(binaryBos);
+      } catch (IOException e) {
+        throw new UncheckedIOException(
+            "Serialization error, this is likely a bug in OpenTelemetry.", e);
+      }
+
+      String json = MarshalerUtil.preserializeJsonFields(realMarshaler);
+
+      cached = new ResourceMarshaler(binaryBos.toByteArray(), json);
```

Comment:
I'm a little weirded out here by the fact that we pre-serialize both the json and the binary, even though we'll only ever be using one of them (in normal circumstances). Is there a way we could only do the one that we need?

Reply:
Yeah this is indeed a bit weird. Since we cache statically, the same marshaler may be doing both binary and json so it needs to hold them both I think (e.g., both logging-otlp and grpc exporters enabled, seems reasonable). I considered an approach with lazy serialization but it seemed to have its own intricacies and preferred the approach that does the classpath check for JSON_AVAILABLE.

I don't think we'll see more singletons added at least for a long time so these two marshalers are very special - maybe it's ok for now?